complex/cargo/list/src/lib.rs

```diff
@@ -37,228 +49,778 @@ pub struct ListManager {
+        if let QueryResults::Rel(rows) = result {
+            if !rows.is_empty() {
+                let row = &rows[0];
+                Ok(Some(Item{
+                    id: row[0].clone().to_inner(),
+                    uuid: row[1].clone().to_inner(),
+                    name: row[2].clone().to_inner(),
+                    due_date: self.fetch_date_for_item("due_date", &uuid).unwrap_or(None),
+                    completion_date: self.fetch_date_for_item("completion_date", &uuid).unwrap_or(None),
+                    labels: self.fetch_labels_for_item(&uuid).unwrap_or(vec![]),
+                }))
+            } else {
+                Ok(None)
             }
+        } else {
+            Ok(None)
         }
-        item_list
     }
 
-    pub fn fetch_item(&self, uuid: &String) -> Option<Item> {
-        let sql = r#"SELECT uuid, name, due_date, completion_date FROM items WHERE uuid=?"#;
-
-        let conn = self.store.read_connection();
-        let mut stmt = conn.prepare(sql).unwrap();
-        let mut item_iter = stmt.query_map(&[uuid], |row| {
-            Item {
-                uuid: row.get(0),
-                name: row.get(1),
-                due_date: row.get(2),
-                completion_date: row.get(3),
-                labels: self.fetch_labels_for_item(uuid)
+    fn fetch_date_for_item(&self, attr: &str, item_id: &Uuid) -> Result<Option<Timespec>, list_errors::Error> {
+        let query = format!(r#"[:find ?{0}
+            :in ?uuid
+            :where
+            [?eid :item/{0} ?{0}]
+            [?eid :item/uuid ?uuid]
+        ]"#, attr);
+        let result = Arc::clone(&self.store).query_args(&query, &[&(&"?uuid".to_string(), &item_id)])?;
+        if let QueryResults::Rel(rows) = result {
+            if !rows.is_empty() {
+                let row = &rows[0];
+                let date: Option<Timespec> = row[0].clone().to_inner();
+                Ok(date)
+            } else {
+                Ok(None)
             }
-        }).unwrap();
-
-        if let Some(result) = item_iter.next() {
-            result.ok()
         } else {
-            println!("No item found for uuid {:?}", uuid);
-            None
+            Ok(None)
         }
+
     }
 
-    pub fn create_item(&self, item: &Item) -> Option<Item> {
-        println!("Creating item {:?}", item);
-        let item_sql = r#"INSERT INTO items (uuid, name, due_date, completion_date) VALUES (?, ?, ?, ?)"#;
-        let conn = self.store.write_connection();
-        let item_uuid = Uuid::new_v4().simple().to_string();
-        conn.execute(item_sql, &[&item_uuid, &item.name, &item.due_date, &item.completion_date]).unwrap();
+    pub fn create_item(&mut self, item: &Item) -> Result<Uuid, list_errors::Error> {
+        let label_str = item.labels.iter().filter_map(|label| {
+            if label.id.is_some() {
+                Some(format!("{}",label.id.to_owned().unwrap().id))
```

Comment:
Firstly, when you're transforming just one leg of a `Result` or `Option,` use `map` and friends.

Secondly, I don't get why you're jumping through `to_owned` and `unwrap`. You should essentially _never_ use `unwrap` anywhere. If you mean `clone`, do that.